Remotion/Web/ClientScript/Scripts/DropDownMenu.ts

```diff
@@ -0,0 +1,928 @@
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+
+interface DropDownMenuResources
+{
+  LoadFailedErrorMessage: string;
+  LoadingStatusMessage: string;
+}
+
+class DropDownMenu_MenuInfo
+{
+  public readonly ID: string;
+  public ItemInfos: Nullable<DropDownMenu_ItemInfo[]>;
+  public readonly LoadMenuItems: (s: (i: DropDownMenu_ItemInfo[]) => void, e: () => void) => void;
+  public readonly Resources: DropDownMenuResources;
+
+  constructor (id: string, loadMenuItems: (s: () => void, e: () => void) => void, resources: DropDownMenuResources)
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsString ('id', id);
+    ArgumentUtility.CheckNotNullAndTypeIsFunction ('loadMenuItems', loadMenuItems);
+    ArgumentUtility.CheckNotNullAndTypeIsObject ('resources', resources);
+
+    this.ID = id;
+    this.ItemInfos = null;
+    this.LoadMenuItems = loadMenuItems;
+    this.Resources = resources;
+  }
+}
+
+class DropDownMenu_ItemInfo
+{
+  constructor (
+      public readonly ID: string,
+      public readonly Category: string,
+      public readonly Text: string,
+      public readonly Icon: Nullable<string>,
+      public readonly IconDisabled: Nullable<string>,
+      public readonly RequiredSelection: number,
+      public IsDisabled: boolean,
+      public readonly Href: Nullable<string>,
+      public readonly Target: Nullable<string>,
+      public readonly DiagnosticMetadata: Nullable<Dictionary<string | boolean>>,
+      public readonly DiagnosticMetadataForCommand: Nullable<Dictionary<string | boolean>>)
+  {
+  }
+}
+
+class DropDownMenu
+{
+  private static _menuInfos: Dictionary<DropDownMenu_MenuInfo> = {};
+  private static _button_timestampDataKey: string = 'clickTimestamp';
+  private static _buttonClassName: string = 'DropDownMenuSelect';
+  private static _itemClassName: string = 'DropDownMenuItem';
+  private static _itemDisabledClassName: string = 'DropDownMenuItemDisabled';
+  private static _itemIconClassName: string = 'DropDownMenuItemIcon';
+  private static _itemSeparatorClassName: string = 'DropDownMenuSeparator';
+  private static _itemSelectedClassName: string = 'selected';
+  private static _focusClassName: string = 'focus';
+  private static _nestedHoverClassName: string = 'nestedHover';
+  private static _currentMenu: Nullable<JQuery> = null;
+  private static _currentPopup: Nullable<HTMLDivElement> = null;
+  private static _currentStatusPopup: Nullable<HTMLDivElement> = null;
+  private static _currentLoadOperation: Nullable<Object> = null;
+  private static _menuItemIDPrefix: string = 'menuItem_';
+  private static _requiredSelectionAny: number = 0;
+  private static _requiredSelectionExactlyOne: number = 1;
+  private static _requiredSelectionOneOrMore: number = 2;
+  private static _itemClickHandler: Nullable<(e: JQueryEventObject) => boolean> = null;
```

Comment:
We should introduce names for any free-floating function type. e.g. `type DropDownMenu_ItemClickHandler = ...`

This simplifies future refactorings if needed. You can assign function types that do not match exactly, for example, one that has less parameters.